### ISSUE TITLE
Add tooling support for injecting collective channels

### ISSUE THREAD

**Issue by benvanik** [iree-org/iree#12113]
After brainstorming with @okkwon we're going to explore synthesizing a `@__set_collective_channels` function that takes a set of `!hal.channel`s as input for use during invocation. The intent is for binding layers (python, pjrt, etc) to use that to inject their own channels that they've coordinated with the orchestration layer (whatever that may be). We can make programs compiled using this work in the various command line tools by adding support in `iree/tooling/context_util.c` or `iree/tooling/device_util.c` for parsing a repeated list of `--channel=` flags and then calling the exported setter to allow the user to specify the topology directly prior to invocation.

Another approach would be to take the channels as arguments on each entry functions. This would make the invocation stateless (besides all the other state we have) such that each invocation within the same context could have a different set of channels, but that too is a special mode unique to certain frontends that we may not want to generalize (it breaks command buffer reuse, etc). In that model instead of having a `--channel=` flag we would have `--input=channel(...)` as a special case that would create them and pass them in along with other arguments. Pending @okkwon's investigation on how this may integrate we can decide exactly which path to take. The setter route would be most canonical with our other uses today so I'm assuming that for now.

--

When going the setter route:

* [ ] Add `--channel=` flag
* [ ] Parse each flag into `iree_hal_channel_params_t` and create a `iree_hal_channel_t`
* [ ] Call `@__set_collective_channels` when available (verify channel count, whether it exists, etc to give good errors)

We'll want to be able to have the queue affinity specified on each flag too for when we have multiple queues backed by different physical devices. The exact format of what we need will depend on some investigations @okkwon is doing as to how best to specify the topology here. We could experiment with placeholders for special values if we need to as well as things that make automation easier to bring up when scripting (allow substitution of environment variables or whatnot).

The other half of this work is synthesizing the `@__set_collective_channels` function in the compiler. This would be similar to what we do during ABI generation where we turn tensor functions into ones taking `!hal.buffer_view` and `!hal.fence`, or create the tflite bindings support functions. That'll likely differ per frontend and live somewhere in the input processing pipeline - how mhlo thinks of channels is weird and specific and not something we can generalize to other frontends (ideally mhlo would just have channels as a concept instead of relying on attributes spread across the IR to implicitly define them).

**Comment by okkwon** [comment]
Thanks for your brain dump. :-) It is a very interesting approach for having a channel as an abstraction between the main process and the IREE code side.

One down side I can think of is the fact that we need to introspect the MLIR code or rely on the code generator side to know what channels we want to pass as a flag. The code produce such as JAX would know the exact details of the channels to pass since it does the data partitioning, but it still needs to do the book keeping or code inspection.

Another difficulty might be the support of dynamic channel creation. This wouldn't be a problem in the near future, but when the program code wants to create a channel dynamically with a group of ranks, the case can't be easily supported by the flag.

Another approach I can think of is to use the host information. This is basically to provide a 1:1 mapping between IP/Port and the rank ID. This seems to be the minimum information to support NCCL channel creation without broadcasting the unique ID of the root process of the group, since every node already knows about the IDs of the other processes. With this approach, we can keep the channel creation logic in the MLIR code and don't need to make the information escape out of the code. (I am assuming that knowing the IP address and port will work for multiple NCCL channel creation, which I believe so.)

Here is an illustration of the case with multiple independent collective ops.

```mlir
// mhlo level
%all_reduce = "mhlo.all_reduce"(%in) ({
  ^bb0(%arg0: tensor<f32> , %arg1: tensor<f32>):
    %add = mhlo.add %arg0, %arg1 : tensor<f32>
    mhlo.return %add : tensor<f32>
  }) {channel_handle = #mhlo.channel_handle<handle = 483, type = 1>, replica_groups = dense<[[0, 1], [2, 3]]> : tensor<2x2xi64>, use_global_device_ids} : (tensor<2304xf32>) -> tensor<2304xf32>
```

The groups [0, 1] and [2, 3] are independent during the communication. The flow level code would look like:

```mlir
// 0, 1, 2, 4 are a global rank ID and there are two independent collective ops with two subgroups.
// membership check
%is_in_0_1 = ... // check the rank ID is in [0, 1]
%is_in_2_3 = ... // check the rank ID is in [2, 3]
// conditional collective operations based on the membership check
scf.if %is_in_0_1
  %channel_0_1 = flow.create_channel group=<[0,1]> affinity=...
  flow.allreduce ..., %channel_0_1
scf.if %is_in_2_3
  %channel_2_3 = flow.create_channel group=<[2,3]> affinity=...
  flow.allreduce ..., %channel_2_3
```

We can further extend the group concept as a separate operation just like MPI does.
```
%cst = arith.const dense<[0, 1]> : tensor<2xi64>
%group_0_1 = flow.create_collective_group %cst -> !hal.collective_group
%channel_0_1 = flow.create_channel_with_group %group_0_1
...
```

In this way, we can support the dynamic group creation. The dynamic aspect will arise when the number of ranks is given as a runtime value.

(Note that I didn't include nccl group start/end calls around the independent communication ops since it is beyond of the discussion here.)

**Comment by okkwon** [comment]
Yeah, the dynamic channel creation scenario is not imminent. The issue will come up when we _want_ to support the number of processes as a dynamic value, similarly to OpenMP where the number of processes is a runtime value.

One critical issue I have is there is no easy way to inject a channel to NCCL. I tried to inject a user provided IP address and port into NCCL using `NCCL_COMM_ID`, but it didn't work; the main reason is NCCL_COMM_ID is used only for the global communicator. To support injecting we need to modify the NCCL source.

If we can create our own repo for a customized NCCL, I'd rather add a NCCL API to create a channel. This would simplify the issue, since the problem is specific to the NCCL environment.

**Comment by benvanik** [comment]
Dynamic channel count is fine - we can pass `!util.list` to functions and fetch the channels we need.

I'm not sure I see any issues here - what we need from an API perspective doesn't change regardless of how channels are created: we just need a simple setter that allows the hosting layer to provide a list of channels to use.

**Comment by okkwon** [comment]
Maybe I think too much including the hosting layer implementation. :)

The second approach, where we specify the information about the channel participants, has a limitation since we can't create a communicator using the vanilla NCCL. Currently the only way to inject `<address>:<port>` into NCCL is `NCCL_COMM_ID`. I tried this and it didn't work well because it is supposed to be used only for creating the global communicator.

The first approach provides a clean separation from the hosting layer and IREE doesn't need to worry about how a channel is created. But it comes with a few things we need to consider.

1. The hosting layer creates a channel. Now the hosting layer also needs to be linked with NCCL to create a communicator. The IREE runtime already uses the NCCL library, so the situation is like two independent layers are using the same NCCL library. Both must use the same NCCL library.
2. The hosting layer needs to inspect the code to collect the channels.
3. Testing becomes difficult since now we need a hosting layer to test. It would be great if we can test our side without deploying any host layer (PJRT or something else).

The root problem that motivated the approaches is that NCCL does not provided a way to broadcast the unique ID to the other processes in the host space. For me, it would be great if NCCL provides a way, just like other communication library, such as MPI. One way is to implement a simple key/value store as an interim solution and treat it as an extension to NCCL but not as a burden for the IREE runtime. (This is mainly to make progress on my side to support the full MHLO ops. We can revisit the first approach as an enhancement step once we have a full PJRT integration.)

**Comment by benvanik** [comment]
I think it's worth the exercise of looking into the NCCL code to demystify things - there's not really any magic in the communicator IDs besides a field literally called `magic` that's just for consistency checking :P

A root in NCCL is (effectively) just a listening socket - all other participants with rank>0 just connect to that socket. `ncclGetUniqueId` returns an automatically selected socket addr (random if, random port), while `NCCL_COMM_ID` lets you specify a socket addr. That's it.

When doing fancy broadcasting/etc as part of bringup all you're doing is _letting each process pick its own socket addr_ (calling `ncclGetUniqueId`) and then exchanging them, but if instead you _assign each process a socket addr_ (setting `NCCL_COMM_ID`) you get the same results: each process you want as a root has an assigned socket addr and is listening on it and other processes know about it and connect.

In the code, `bootstrapNetInit` is what either picks its own socket addr or parses the `NCCL_COMM_ID` provided - the result is stored in `bootstrapNetIfAddr`:
https://sourcegraph.com/github.com/NVIDIA/nccl@f3d51667838f7542df8ea32ea4e144d812b3ed7c/-/blob/src/bootstrap.cc?L30-54

The code paths then diverge a bit:

If letting NCCL self-select an address when you call `ncclGetUniqueId` it uses that `bootstrapNetIfAddr` from above (which may be from `NCCL_COMM_ID`!) and calls `bootstrapCreateRoot`, which is the thing that actually listens on the socket (the most important part):
https://sourcegraph.com/github.com/NVIDIA/nccl@f3d51667838f7542df8ea32ea4e144d812b3ed7c/-/blob/src/bootstrap.cc?L200-201

If specifying an address then `ncclGetUniqueId` just turns into an identity using the specified address and instead it is `ncclCommInitRankDev` that calls `bootstrapCreateRoot` when `myrank == 0`:
https://sourcegraph.com/github.com/NVIDIA/nccl@f3d5166/-/blob/src/init.cc?L1142-1145&subtree=true

In both cases the only load-bearing thing is that `bootstrapCreateRoot` is called with the local address and it creates the listening socket.

It'd of course be nice if there was a NCCL API for saying `bind to this address` (`ncclSetUniqueId`), but in lieu of that using `setenv` to set `NCCL_COMM_ID`, calling `ncclGetUniqueId` to have NCCL parse that/generate the `magic`, and then calling `ncclCommInitRankDev` once gives you precise control over the creation of the listening socket that's required.

Now the trick comes in with the `magic`: that's the only thing that is not _easily_ accessible. But only somewhat: AFAICT it's only used for sanity checking and the value is semantically meaningless (https://sourcegraph.com/github.com/NVIDIA/nccl@f3d5166/-/blob/src/misc/socket.cc?L427:20&subtree=true). This is where a C API would help but otherwise the NCCL IDs are well-formed and we can poke them:
```
struct ncclBootstrapHandle {
  uint64_t magic;
  union ncclSocketAddress addr;
};
static_assert(sizeof(struct ncclBootstrapHandle) <= sizeof(ncclUniqueId), "Bootstrap handle is too large to fit inside NCCL unique ID");
```
So if we call `ncclGetUniqueId` to parse an address and get back random magic we can then just set the magic to 0/whatever and we have a valid deterministic `ncclUniqueId` we can use. No need to exchange peer-to-peer after the fact as prior to any child processes being launched we can control the `ncclUniqueId` of all of them.

---

Not saying we actually do that (poke `magic`) but for basic test orchestration it would be enough - whatever is launching the processes knows the address of where each process is being launched (in testing likely all localhost) and can pick whatever ports it wants. Normal bsd sockets stuff - either limit to one process cluster at a time and use fixed ports or do bog standard socket probing to find open port ranges in the orchestration layer. Each process gets a list of all the other addresses/ports as args and can then open/bind communicators as needed. This would work on the command line just fine and you can have a bash script that hardcodes the ports and reproducibly brings up clusters. Sure it won't scale across a datacenter but we aren't trying to scale IREE's testing infra like that anyway - that's for the layers above IREE to do - this is about finding a way to test all this stuff without needing to integrate large projects/build out lots of infrastructure. Testing clusters of network services is classically done like this, and hopefully you can see that all we're doing (in the complex case of multiple roots) is setting up a peer-to-peer topology network service.

Beyond that there's a large spectrum between that simple approach and key-value stores and other heavyweight stuff. Named pipes and such are usually used for things like this in smaller settings (local machine or NFS) and work with (practically) just fopen/fread/fwrite. Moving one step beyond that bsd sockets has zero deps and would let us exchange over remote machines with a single orchestrator address (all children connect to the assigned root and post their ncclUniqueId + read back all the others). It's what NCCL is doing itself so it's not introducing any requirements NCCL doesn't already have.

But I think if wanting to build something more "real" instead of just for testing then we should just _use NCCL itself_ to exchange addresses: the orchestrator can be the root and all children can participate in an all-gather. This doesn't mean the orchestrator needs to link in NCCL and call it directly - _it can use the HAL_! This would let it work across any backend we have with collectives support (CPU/ROCM/etc) as if we can all-gather a buffer of user data we can also all-gather a buffer of addresses. Interestingly, NCCL uses itself to do some of this kind of work during startup (https://sourcegraph.com/github.com/NVIDIA/nccl@f3d5166/-/blob/src/init.cc?L574&subtree=true). In other words, instead of using MPI or whatnot to pass things around we can use NCCL in the same way via the HAL.

So the orchestrator could create a HAL device, create a channel using `iree_hal_channel_create` as root (rank=0, count=child count), query its own channel ID (we can have an `iree_hal_channel_query_id(iree_byte_span_t out_buffer)` method to return it), launch all the child processes with a flag set to the queried ID, and submit a command buffer with `iree_hal_command_buffer_collective(...IREE_HAL_COLLECTIVE_KIND_ALL_GATHER...)` - each child process would do the same. Now all processes (including the orchestrator) are connected and have each other's NCCL IDs as generated from NCCL itself and that can be used to create any number of other channel groups. This whole thing could be wrapped up in `iree/tooling/collective_util.h` to make it turn-key in our tools (iree-run-module/iree-benchmark-module/etc). In this model `NCCL_COMM_ID` is only used for the orchestrator ID and there are zero additional deps. The orchestrator could be in python or whatever calling through the runtime python bindings, though a simple C tool would be easiest.

---

So now if we have a way for all processes to get all other processes' root IDs we can create arbitrary subgroups as needed by emitting the `__set_collective_channels` function or something like it. The easiest thing would be to pass in the full set of coordination channels such that we could completely control the behavior in the model itself with an `__init_collective_channels(%self_id: i32, %all_participants: !util.list<!hal.channel>)`. New channels could be created if we add the `hal.channel.id` query and then implement the TODO for letting us pass that ID on creation: https://github.com/google/iree/blob/5fa81548dd01941ecc3487ed37c36343c4ccf178/compiler/src/iree/compiler/Dialect/HAL/IR/HALOps.td/#L699). This turns the channels we've setup into (expensive) sockets: the code we emit can send/recv/etc to individual participants in whatever way they want. Could look something like:
```mlir
util.global private @group_1_2 : !hal.channel
func.func @__init_collective_channels(%self_id: i32, %all_participants: !util.list<!hal.channel>) {
  // form a group with participants 1,2, only participants 1 and 2 do this and 1 becomes the root:
  scf.if %self_id == 1 || %self_id == 2 ... {
    %p1 = util.list.get %all_participants[%c1] : !util.list<!hal.channel>
    // ... exchange IDs by all-gather on nodes 1/2
    %group_1_2 = hal.channel.create %p1_id ... rank=%sub_rank : !hal.channel
    util.global.store %group_1_2, @group_1_2 : !hal.channel
  }
}
```
...but it's just a program and it can do whatever kind of configuration it wants.

The only thing the user does in this case is get the processes participating together and everything else is controlled by the program we emit so no need for side-channel metadata and such.

Now when it comes to hosting layers they can do whatever they want but will bottom out on this. If they want to use MPI to exchange IDs/addresses, or use key-value stores, or protobufs over gRPC, or whatever they can. Instead of using our helper they can call `iree_hal_channel_create` with params derived in whatever way they can for each participant.

Knowing the exact calling behavior from PJRT would help us figure out a command line tool that could do this for us - I'm imagining `iree-launch-multi-device --device=cuda://0 --device=cuda://1 -- iree-run-module --module=module.vmfb --input=....` or something that would play the role of orchestrator by creating the root channel, forking off the post-`--` binary with the appropriate `--device=` and root channel ID, and then wait for it to complete - not sure if we want to include sharding of arguments and such in that or not though. Or an `iree-launch-cluster` that would take addresses and an `iree-cluster-server` that would listen and fork as needed for multi-host stuff but with effectively the same behavior (launch-cluster would connect to each daemon, send the params, and then they'd all-gather using the launcher address).

I'd estimate it'd take me a day or two to build something like those launchers (I've written 5 iree-* tools and have become exceedingly efficient at it lol), but if we could specify how we wanted them to behave that'd make it easier to ensure it worked for what you needed. We could also then agree on the spec of the function we emit in the compiler for constructing the groups - to start it could just take one channel and use it in place of the default one.

**Comment by okkwon** [comment]
Yes, I did some more exploration about the NCCL calls over the weekend since my initial attempt to workaround to get the uniqueID from `NCCL_COMM_ID`. I injected a list of address and port pairs and created uniqued IDs in advance, but it didn't work as I expected; 
`ncclCommInitRankDev()` didn't call `bootstrapCreateRoot()` because when it is called `NCCL_COMM_ID` is cleared.

So when the socket is open depends on the present of NCCL_COMM_ID. When it is used, a socket is open in `ncclCommInitRankDev()`, but for the other case, it is open in `ncclGetUniqueId()`.

We may be able to find a solution by calling those two functions with the same NCCL_COMM_ID, but my debugging experience taught me that this can be very hacky knowing that the behavior is really the NCCL's implementation detail. What if the NCCL implementation is changed?

When I searched through the NCCL github issue pages for the right usage of the NCCL_COMM_ID, I learned that its usage is for the global communicator. A user reported that this can't be used for the cases where nodes are in different subnets due to another quirk in implementation details.

I am inclined to a distributed key/value store since its implementation cost is very low. I expect I can implement and integrate one to the IREE CUDA NCCL implementation in a week or two. This also works for all the systems with multiple nodes. So, this can be used for a single-host and multi-host usages. So until we start digging the performance with a real system for a customer use case, I think it can serve us well.

The user command will be like:
```
kvs_server <port>  # launching in the server
iree-run-module --kvs=<addr>:<port> ... &
iree-run-module --kvs=<addr>:<port> ... &
```
or using an environmental variable
```
export IREE_CUDA_NCCL_KVS=<addr>:<port>
...commands...
```

We can also make the root host to launch it as a service during the runtime.

The client interface to the store is like
1. connect(addr, port)
2. get(key)
3. set(key, value)

This can be made pluggable by using a predefined name for a dynamic library, e.g., `lib-kvs.so, which implements the above functions. The implementation can be a user choice. It can be a shared file in NFS, redis, grpc, socket, MPI, or so on.

This is how the Gloo project made the abstraction.

In summary, I think the key/value store interface is simple, extensible, and cheap to implement (one or two weeks) to augment the NCCL's missing part.

Here is a feature table:

| Solution | coverage | performance | implementation cost | Notes |
|----------|----------|-------------|---------------------|-------|
| MPI      | most but not google infra! | high | medium (dlopen)| Nice to have as a debugging tool |
| Key/Value interface | all | good or high | low (2 weeks with a gRPC impl.) |  |
| shared file| single host | low | low (no external deps) | perf varies on FS perf. |
| NCCL hack | all | high | high (need to define what the hosting layer does.) | Fragile |
| socket | all | high | can be tricky to support the all environments with different interconnect medium |   |

Once we have a full feature to support TF and JAX use cases, we should definitely revisit this part as performance investigation.

------------------

Regarding the multi process launcher, there are some tools we may reuse: mpirun or GNU parallel. There may be more but we can start with these tools first. (mpirun doesn't require the program is built with an MPI library so we can just use it as a launcher program.)

Thanks again for the comments! It is really amusing to see you dig so much in the NCCL library. :) :p

**Comment by allieculp** [comment]
@benvanik @okkwon Is this still active? Please bump down to P2 if needed.

**Comment by okkwon** [comment]
The issue is resolved by the channel provider method. I marked it as P3. @benvanik let us know if there is any remaining task or issue you want to tackle from this issue. Otherwise, please close it. Thanks!

**Comment by benvanik** [comment]
yeah I think we can call it closed now! the way we ended up with is way nicer than what I originally was thinking (though we may still need something like that for certain frontends where the default channel needs to be controlled). the provider API will need some refactoring be we can track that separately as needed.
